liquid-compiler/src/parser.rs

```diff
@@ -198,85 +188,84 @@ fn parse_tag(
-                    }
-                    if nesting_depth == 0 {
-                        break;
-                    }
-                }
-
-                // Now that we know this element won't close this block,
-                // add it to the vector.
-                block.push(element);
-        }
-        options.blocks[name].parse(name, tokens, block, options)
-}
 
 /// An interface to parse elements inside blocks without exposing the Pair structures
 // TODO find if there is a better way to store the blocks instead of a vector of pairs.
-pub struct TagBlock<'a>(Vec<Pair<'a>>);
-impl<'a> TagBlock<'a> {
-    fn new() -> Self {
-        TagBlock(Vec::new())
+pub struct TagBlock<'a:'b, 'b> {
+    name: &'b str,
+    iter: &'b mut Iterator<Item = Pair<'a>>,
+    nesting_depth: u32,
+}
+
+impl<'a, 'b> TagBlock<'a, 'b> {
+    fn new(name: &'b str, next_elements: &'b mut Iterator<Item = Pair<'a>>) -> Self {
+        TagBlock {
+            name,
+            iter: next_elements,
+            nesting_depth: 1,
+        }
     }
 
-    fn push(&mut self, element: Pair<'a>) {
-        self.0.push(element);
+    fn next(&mut self) -> Result<Option<Pair<'a>>> {
+        if self.nesting_depth == 0 {
+            return Ok(None);
+        }
+
+        let element = self
+            .iter
+            .next()
+            .expect("The file must not end before an EOI.");
+        if element.as_rule() == Rule::EOI {
+            return panic!("Errors not implemented. Unclosed block.");
+        }
+        if element.as_rule() == Rule::Tag {
+            let nested_tag_name = element
+                .clone() // Maybe there is a better way?
+                .into_inner()
+                .next()
+                .expect("Tags start by their identifier.")
+                .as_str();
+            if self.name == nested_tag_name {
+                self.nesting_depth += 1;
+            } else if format!("end{}", self.name) == nested_tag_name {
+                self.nesting_depth -= 1;
+            }
+            if self.nesting_depth == 0 {
+                return Ok(None);
+            }
+        }
+        Ok(Some(element))
     }
 
-    pub fn parse(self, options: &LiquidOptions) -> Result<Vec<Box<Renderable>>> {
+    pub fn parse(mut self, options: &LiquidOptions) -> Result<Vec<Box<Renderable>>> {
         let mut renderables = Vec::new();
-        let mut pairs = self.0.into_iter();
-        while let Some(element) = pairs.next() {
-            renderables.push(parse_element(element, &mut pairs, options)?);
+        while let Some(element) = self.next()? {
+            renderables.push(parse_element(element, self.iter, options)?);
         }
         Ok(renderables)
     }
-    
+
     // Probably a to_str might be implemented by checking the start of 1st pair and end of the last.
     // Then just create a slice directly from the input text.
-    pub fn to_string(&self) -> String {
+    pub fn to_string(&mut self) -> Result<String> {
```

Comment:
Should this just take ownership at this point?

Reply:
The problem with taking ownership of `TagBlock` is that it would be unavailable to call `close()` at the end of `parse_tag()`. So the `close()` function is the one that currently can take ownership.

Unless we find a better approach than the `close()` function to ensure the block is parsed to its end, I don't think we can change this.

Reply:
Regarding the first thought: the TagBlock never exposes Pest's API (and thus `next()`). As such, there is no way for `raw` to do this, as the only function that would be available besides `to_string` is `parse`, which would actually parse the elements instead of returning them as a string. There may also be other tags (eventually) that use this function (or something similar) if they need something different than just directly parsing the elements of the block.

On the second thought: that might be a possibility, although I'm not really sure if there is a way to do the error check without still needing something like the `close()` function (and thus a `&mut`). Do you have any ideas?

Reply:
I'm marking this as resolved, as I think that if something is to change about the way `close()` works, it's better to do it in a future PR.

Regarding the other thought, it is already solved.

Reply:
Actually, as I was writting this I remembered about the `Drop` trait.
As it may be a small change, I will see if it can be used to fix this in this PR. Otherwise, I'll let this for the future.

Reply:
Ok, this is now solved and `TagBlock` is no longer passed to `block` functions as a `&mut`.